5주차 미션 제출 (이선환)

glowing713 opened the pull request:
# 5주차 미션 제출 (이선환)

## 접근 방법

- 이번 프로젝트는 실제 서비스 중인 쇼핑몰 앱을 참고하여 구현하였습니다.
- 개발에 앞서서 장바구니API가 어떤 형태의 데이터를 불러오게 되는지, 어떤 컴포넌트들을 사용하는지, 각 컴포넌트는 어떤 형식으로 데이터를 주고 받는지(props, store) 설계하는 시간을 충분히 갖고, 문서화를 진행했습니다.

## 특이사항

- 초반에 많은 기능을 담으려고 하다보니 시간에 쫓기는 감이 없지않아 있었던 것 같습니다ㅠㅠ 프로젝트 설계 단계에서 작업량을 잘 체크하고 진행해야겠다는 생각이 들었습니다.

## 제출 후 확인

* [ ] 본인의 구현 결과를 `dev` 브랜치에 구현하고, `dev` 브랜치에서 `submission` 브랜치로 PR을 보내주세요.

CC: @externship-master  @lkaybob

glowing713 committed:
feat: Add CartPage component and routing

glowing713 committed:
test: Add cart page link routing test

glowing713 committed:
feat: Add cart page container

glowing713 committed:
chore: Add data test property to cartpage link

glowing713 committed:
test: Test header container rendering

glowing713 committed:
feat: Add header container

glowing713 committed:
test: Test goback button rendering

glowing713 committed:
test: Test redirection to previous page

glowing713 committed:
feat: Add event to go back button

glowing713 committed:
chore: Delete mock implementation(not using)

glowing713 committed:
test: Test page title rendering

glowing713 committed:
feat: Add page title

glowing713 committed:
refactor: Add CartHeader component

glowing713 committed:
test: Test sticky header in cartpage content container

glowing713 committed:
feat: Implement sticky header in cartpage

glowing713 committed:
style: Add style to cart content sticky header

glowing713 committed:
refactor: Add content header component

glowing713 committed:
chore: Delete contentheader component tc

glowing713 committed:
test: Test sticky footer rendering

glowing713 committed:
feat: Add sticky footer component

glowing713 committed:
test: Test order button rendering in sticky footer

glowing713 committed:
feat: Add order btn in sticky footer

glowing713 committed:
style: Add style to order button

glowing713 committed:
test: Test setCartItem mutation commit

glowing713 committed:
feat: Add cartitem state and mutations in store

glowing713 committed:
chore: Fix style

glowing713 committed:
feat: Add cart product container title

glowing713 committed:
test: Add CartItem component test code with propData

glowing713 committed:
feat: Add CartItem component with prop data

glowing713 committed:
test: Test checkbox rendering

glowing713 committed:
feat: Add checkbox in CartItem

glowing713 committed:
feat: Add props to CartItem component and test code

glowing713 committed:
test: Test product image rendering

glowing713 committed:
test: Test product name rendering

glowing713 committed:
feat: Add product name in cartitem

glowing713 committed:
test: Test product description rendering

glowing713 committed:
feat: Add product description in cartitem

glowing713 committed:
test: Test product price rendering

glowing713 committed:
feat: Add product price in cartitem

glowing713 committed:
test: Test conditional price rendering

glowing713 committed:
feat: Change original price to be rendered on sale

glowing713 committed:
test: Test delete button and emit in cartitem
삭제버튼을 클릭하면 부모 컴포넌트에서 해당 아이템을 삭제하는 이벤트를 emit 한다.

glowing713 committed:
feat: Add delete button and emit method

glowing713 committed:
test: Test delCartItem mutation on delete-item event trigger

glowing713 committed:
feat: Add delete-item custum event handler

glowing713 committed:
feat: Add number formatting to prices

glowing713 committed:
feat: Add reactive cart item count in header

glowing713 committed:
feat: Calculate total price and pass to footer

glowing713 committed:
chore: Refactor expressions in template to computed property

glowing713 committed:
feat: Add total price container

glowing713 committed:
style: Add style

glowing713 committed:
chore: Refactor expression in template to computed property

glowing713 committed:
feat: Add order component

glowing713 committed:
feat: Add order complete page

glowing713 committed:
chore: Fix indentation

glowing713 committed:
chore: Refactor expression in template to computed

glowing713 commented:
시간관계상 PR을 자세히 작성하지 못했습니다ㅠㅠ 리뷰영상에 제가 작성한 코드의 주요 설명이 있으니 영상을 참고해주시면서 코드리뷰 진행해주시면 감사하겠습니다..! 5주차까지 정말정말 고생많으셨습니다!!!😃

k12ms26 commented on mission/src/router/routes.js:
```diff
@@ -0,0 +1,35 @@
+import CartOrder from '@/views/CartOrder.vue';
+import OrderComplete from '@/views/OrderComplete.vue';
+
+const routes = [
```
routes 파일을 따로 빼서 index.js에 선언한 이유가 있으실까요??

glowing713 replied:
아! 이부분은 강사님 코드를 참고하면서 반영한 수정사항인데요,
이렇게 라우트를 정의하는 부분과 이를 불러와서 라우터 객체를 선언하는 부분을 분리시키면
추후에 애플리케이션 규모가 확장되는 상황에서 더 코드가 깔끔하지 않을까 하여 적용해보았습니다:)

k12ms26 replied:
@glowing713 아하 그렇군요 그렇다면 파일명에 이 라우터 파일이 어떤 용도인지 명시해주시면 이해가 더 쉬울것같기도합니다!! 답변 감사드려요 😄

glowing713 replied:
넵!! 좀더 명시적인 파일명으로 바꿔보겠습니다!ㅎㅎ 감사합니다 민선님:)

k12ms26 reviewed (commented):
[좋은 점/잘한 점]
실제 쇼핑몰처럼 구현하셔서 UI도 깔끔하고 작성하신 코드도 전체적으로 깔끔하다고 생각했습니다! 전체선택 부분 (ContentHeader) 까지 유닛 테스트하신것이 인상깊었습니다 😃 

[의견제시]
코멘트 달아두었습니다 :) 그동안 너무 고생많으셨어요 😄

ohyein00 commented:
- 이번 주차에 저는 아직 테스트를 하지 못했음에도 매우 막막했는데 
선환님께서 테스트를 굉장히 꼼꼼하게 잘 짜주셔서 많은 참고가 될 것 같습니다👍 감사합니다
- 실제 서비스중인 앱을 참고해주셨다고 해서 더욱 인상 깊게 보았습니다. 
여러 기능들도 꼼꼼히 구현하셔서 실제 쇼핑몰 같아요!
- 그동안 고생 많으셨습니다. 5주차동안 선환님의 코드에서 많은 것들을 배우고 갔어요 감사합니다🥰

ohyein00 commented on mission/src/components/CartPage/CartItem.vue:
```diff
@@ -40,6 +43,11 @@ export default {
       return !(this.price === this.originalPrice);
     },
   },
+  methods: {
+    emitDeleteItem() {
+      this.$emit('deleteItem', this.productNo);
+    },
+  },
```
해당 컴포넌트에서 바로 commit을 하지 않고 커스텀 이벤트를 사용하시게 된 이유가 궁금합니다!

glowing713 replied:
엏..! 그러고보니 저 컴포넌트 내부에서 바로 수정하는 방법도 있었겠군요..!!
장바구니 상품 추가/제거와 같은 로직은 부모컴포넌트(CartPage)에서만 처리하도록 하고싶어서 저렇게 구현했었습니다!
설계단계에서 저렇게 생각을 하고 개발을 진행하다보니 자식컴포넌트에서 바로 상태를 업데이트할 생각은 못했었네요😅
해당 컴포넌트에서 바로 commit을 해주는게 좀 더 나을까요..?

k12ms26 replied:
@ohyein00 @glowing713 저는 emit하는게 뭔가 더 부모-자식 컴포넌트 규칙을 더 잘 이해하는 느낌이 드는 코드 같네요..! 뭐가 더 좋은지 저도 궁금합니다!!

jambottle replied:
만약 이게 부모-자식 관계 안에서 일어나는 동작이라면 민선님(@k12ms26) 말씀처럼 `$emit`을 쓰는 게 맞는 것 같고,
여기서 `deleteItem`이라는 메서드가 Vuex Store에 저장되어 있는 데이터에 접근하고 있는 용도라면 예인님(@ohyein00) 질문처럼 굳이 `$emit`으로 여러 컴포넌트의 통신을 거치는 것이 아니라 바로 Store에 접근하는 게 Vuex의 의도상 깔끔하지 않을까요...? 🤔

lkaybob replied:
@jambottle 님이 말씀해주신 것처럼 Vuex 의도상 따로 event를 타는 것보다는 store의 action 등을 통해 변경을 불러내는 것이 더 좋지 않을까 합니다. '삭제'에 대한 action이 store에서 정의됐다면 굳이 다른 컴포넌트를 통해서 action을 수행하는 것은 오히려 데이터 흐름을 이해하는데 혼동을 줄 수 있지 않을까 합니다.

glowing713 replied:
@ohyein00 @k12ms26 @jambottle @lkaybob 피드백 감사합니다😃 말씀해주신 것처럼 Vuex를 사용하는만큼 별도의 emit 없이 바로 Store에 접근하는 방법이 여러 측면에서 좋아보입니다..! 바로 반영해보겠습니다~!🙏🏻

daaeaad reviewed (commented):
- 작업 시작 전에 실제 앱을 참고하여 기획문서를 작성하시고 노션으로 작업 내용을 잘 정리하신 점이 돋보입니다 :)
- 저도 초반에 욕심이 앞서 시간에 많이 쫓겨버렸는데 그런 점을 언급해주셔서 공감이 많이 되었습니다 ㅠㅠ 그럼에도 불구하고 필요한 기능 구현뿐만 아니라 각 영역을 컴포넌트로 잘 분리하고 테스트 코드도 작성하신 점에서 선환님이 최선을 다하셨다는걸 느낄 수 있었어요 :)

**[ 의견 ]**
- 아직 의견을 남길만한 부분을 찾지 못했지만 추후 남길 의견이 생각나면 다시 오도록 하겠습니다..!
- 5주동안 고생 많으셨어요 :)

Subinhyun reviewed (commented):
실제 서비스 중인 앱을 참고하여 구현해주셔서 진짜 쇼핑몰 같았어요!
추가로 구현하신 부분 참고해서 저도 보완해야 될 것 같습니다 ㅜ
전체적으로 코드가 깔끔해서 좋았습니다!! 
저는 TC 작성을 하지 못했는데 선환님 코드를 참고해야겠네요 ㅠㅠ 
수고 많으셨습니다 !

ghost commented:
안녕하세요 선환님! 남겨주신 영상 잘 보았습니다! 커밋 메세지만 보아도 얼마나 열정적으로 미션에 임하셨는지 알 수 있었습니다. 그만큼 결과물도 실제 서비스중인 쇼핑몰처럼 훌륭한 것 같고 굉장히 사용하기 편할 것 같다는 느낌이 듭니다.! 추가적으로 구현해주신 전체 선택, 선택 삭제 기능들도 인상깊어 저도 참고하여 구현해보고 싶었습니다..! TC도 견고하게 작성하신 것 같아 배울점이 많은 것 같습니다. 5주동안 고생많으셨습니다!

jambottle commented on mission/src/App.vue:
```diff
@@ -3,13 +3,18 @@
 
 <style>
-html, body {
+html,
+body {
   width: 100%;
   height: 100%;
   margin: 0;
   padding: 0;
 }
 
+a, a :visited {
+  color: inherit;
+}
```
이 부분을 보니까 `reset.css`의 템플릿이나 `common.css`의 방법론을 찾아보시면,
선환님이 의도하고 계신 목표에 부합하는 결과를 만나실 수 있을 것 같아요오-

glowing713 replied:
오,, 안그래도 이번 익스턴십에서 다른분들 코드를 보면서 css 초기화하는 템플릿을 많이 접할 수 있었는데,
저도 꼭 적용해보아야겠습니다😃 감사합니다!

jambottle commented on mission/src/components/CartPage/CartItem.vue:
```diff
@@ -0,0 +1,158 @@
+<script>
+export default {
+  name: 'CartItem',
+  props: {
+    productNo: { type: String, default: '' },
+    name: { type: String, default: '' },
+    image: { type: String, default: '' },
+    price: { type: Number, default: 0 },
+    originalPrice: { type: Number, default: 0 },
+    description: { type: String, default: '' },
+  },
```
일일이 props/data마다 타입과 기본값 지정해주시는 습관 좋은 것 같아요 👍

jambottle commented on mission/src/components/CartPage/CartItem.vue:
```diff
@@ -0,0 +1,158 @@
+  text-decoration: line-through;
+}
+
+@media (min-width: 390px) {
+  .item-info img {
+    max-width: 90px;
+  }
+}
+
+@media (min-width: 820px) {
+  .item-info img {
+    max-width: 100px;
+  }
+}
+
+@media (min-width: 1024px) {
+  .item-info img {
+    max-width: 130px;
+  }
```
반응형..!! ㄴ😲ㄱ

2taesung replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 재원님 ㅋㅋㅋ 리액션 굿

jambottle commented on mission/src/views/CartPage.vue:
```diff
@@ -0,0 +1,154 @@
+    };
+  },
+  computed: {
+    cartItems() {
+      return this.$store.state.cartItems;
+    },
```
만일 state를 더 많이 가져와야 하는 경우에는 `mapState` 헬퍼로 훨씬 쉽고 편하게 state들을 가져올 수 있어요!!

**🔍참고**(Vuex 공식문서): https://vuex.vuejs.org/guide/getters.html#the-mapgetters-helper

glowing713 replied:
이번 5주차 재원님 코드보면서 헬퍼함수 적극적으로 사용해봐야겠다는 생각이 들었습니다!🔥 감사합니다:D

jambottle commented on mission/src/views/CartPage.vue:
```diff
@@ -0,0 +1,154 @@
+      return this.cartItems[0].description;
+    },
+  },
+  methods: {
+    removeItem(id) {
+      this.$store.commit('delCartItem', id);
+    },
+    setItems() {
+      this.$store.commit('setCartItems');
+    },
+  },
```
여기서도 `mapMutations` 헬퍼를 활용하면, 훨씬 쉽고 편하게 가져올 수 있어요!!

**🔍참고**(Vuex 공식문서): https://vuex.vuejs.org/guide/mutations.html#committing-mutations-in-components

jambottle reviewed (approved):
선환님, 안녕하세요 👋 어느덧 이번이 마지막 과제 리뷰 시간이네요오-
여태껏 같은 과정의 일원으로 함께할 수 있어서 즐거웠고, 아래와 같이 마지막 피어 리뷰를 남깁니다!

> 강사님이 사전에 제공해주신 API를 사용하지는 않으셨지만, **Vuex Store에 여러 종류의 상품을 Mock Data로 담아 실제 있는 사이트처럼 완성도 높은 페이지를 구현하신 것이 인상적이었어요!** 내부에 어떤 로직을 구현해 작동시키느냐도 중요하지만, 그 로직을 바탕으로 보이는 화면 부분도 중요하기 때문에 이번 미션을 통해서 그런 부분들을 잘 만족시키신 것 같고, **특히 ABLY를 직접 참고하신 적극적인 모습이 보기 좋았습니다 👏👏👏**

> 프로젝트에 대한 사전 조사와 더불어, 해당 내용들을 바탕으로 기획할 부분을 선정하고, 그 내용을 직접 구현하기 위한 테스트 코드 작성까지... 일련의 과정이 수준 높은 개발을 완성하시는 데에 많은 도움이 됐을 것 같습니다. 처음 피어 리뷰로 뵙지만, 덕분에 많이 배워갑니다!! 🙏

그동안 고생 많으셨고, 앞으로도 지금처럼 파이팅하시길 기원하겠습니다 🙌

2taesung commented:
선환님 리뷰를 할 때면 미션 기본 요구사항 뿐 아니라 실제 쇼핑몰 서비스를 개발하는 듯한 느낌을 받았습니다.
큰 동기부여도 받아갑니다! 선환님이 고려하신 부분들을 꼭 짚어봐야겠네요 
5주동안 수고 많으셨습니다!

lkaybob commented on mission/src/components/CartPage/ContentFooter.vue:
```diff
@@ -0,0 +1,52 @@
+    goToOrder() {
+      this.$router.push({
+        name: 'CartOrder',
+        params: {
```
장바구니에서 파악할 수 있는 데이터들에 대해서는 route의 param을 타지 않고 store에서 가져올 수 있습니다.
* 라우팅은 해당 컴포넌트로 이동되도록 수정하시고, CartOrder.vue에서는 store에 getter들을 추가해 구현해보셨으면 합니다.

glowing713 replied:
넵~! store를 더 활용하는 코드로 수정해보겠습니다:-)

lkaybob commented on mission/src/store/index.js:
```diff
@@ -1,12 +1,16 @@
   state: {
+    cartItems: [],
   },
   mutations: {
```
route.param에 실어보냈던 정보들은 여기서 getter를 정의해 사용해볼 수 있습니다.

lkaybob reviewed (commented):
(코드 리뷰가 누락된 것을 뒤늦게 확인했습니다. 뒤늦게나마 코멘트 남겨드립니다. 늦어서 죄송합니다.)

주어진 요구사항을 구현하기 위해 나름의 생각을 잘 정리하시고 이를 잘 구현해주신 것 같습니다. 각 UI 요소들을 규칙을 가지고 컴포넌트화해주신 부분도 돋보이고, 이전에 알려드린 내용들을 최대한 반영해서 구현해주시려는 모습도 돋보였습니다.

디테일한 구현 측면에서 다소 아쉬운 부분이 존재했는데, 라인 코멘트 확인하시고 개선해보시면 좋을 것 같습니다.

5주동안 고생하셨습니다.

glowing713 committed:
refactor: Rename route object file

glowing713 committed:
refactor: Move state mutation from cartpage to cartitem
기존에는 삭제 버튼을 클릭할 경우, 부모 컴포넌트로 emit하는 부분이 있었으나,
불필요한 흐름으로 판단이 되어, 해당 컴포넌트인 CartItem에서 바로
store mutation을 commit하도록 변경.

glowing713 committed:
refactor: Use store helper
헬퍼 함수를 사용하여 store 접근하는 부분을 수정

glowing713 committed:
feat: Get checked items with v-model in CartPage compo
CartPage 컴포넌트에서 modelValue prop에 checked를 바인딩하고,
CartItem 컴포넌트에서는 checkbox 상태에 따라 modelValue prop으로 받은 배열을
업데이트해서 update:model-value 이벤트를 새로운 배열과 함께 emit한다.
CartPage 컴포넌트에서 새롭게 받은 배열을 checked 배열에 할당하여 업데이트한다.

glowing713 committed:
chore: Add vetur formatting

glowing713 committed:
feat: Add select/deselect all button
v-model을 사용하여 전체선택/해제 기능 구현
